cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -429,6 +429,169 @@ const FunctionDoc make_struct_doc{"Wrap Arrays into a StructArray",
+    }
+    for (int64_t map_array_idx = 0; map_array_idx < map_array.length(); ++map_array_idx) {
+      if (!map_array.IsValid(map_array_idx)) {
+        RETURN_NOT_OK(builder->AppendNull());
+        continue;
+      } else {
+        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> scalar,
+                              map_array.GetScalar(map_array_idx));
+        auto map_scalar = std::static_pointer_cast<MapScalar>(scalar);
+        ARROW_ASSIGN_OR_RAISE(auto scalar_output, GetScalarOutput(ctx, *map_scalar));
+        RETURN_NOT_OK(builder->AppendScalar(*scalar_output));
+      }
+    }
+    ARROW_ASSIGN_OR_RAISE(auto result, builder->Finish());
+    out->value = result->data();
+    return Status::OK();
+  }
+
+  static Status ExecMapScalar(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const auto& options = OptionsWrapper<MapArrayLookupOptions>::Get(ctx);
+    std::shared_ptr<DataType> item_type =
+        checked_cast<const MapType&>(*batch[0].type()).item_type();
+
+    const auto& map_scalar = batch[0].scalar_as<MapScalar>();
+
+    if (ARROW_PREDICT_FALSE(!map_scalar.is_valid)) {
+      if (options.occurrence == MapArrayLookupOptions::Occurrence::ALL) {
+        out->value = MakeNullScalar(list(item_type));
+      } else {
+        out->value = MakeNullScalar(item_type);
+      }
+      return Status::OK();
+    }
+
+    ARROW_ASSIGN_OR_RAISE(out->value, GetScalarOutput(ctx, map_scalar));
+    return Status::OK();
+  }
+};
+
+Result<ValueDescr> ResolveMapArrayLookupType(KernelContext* ctx,
+                                             const std::vector<ValueDescr>& descrs) {
+  const auto& options = OptionsWrapper<MapArrayLookupOptions>::Get(ctx);
+  std::shared_ptr<DataType> type = descrs.front().type;
+  std::shared_ptr<DataType> item_type = checked_cast<const MapType&>(*type).item_type();
+  std::shared_ptr<DataType> key_type = checked_cast<const MapType&>(*type).key_type();
+
+  if (!options.query_key || !options.query_key->type ||
+      !options.query_key->type->Equals(key_type)) {
+    return Status::TypeError(
+        "map_array_lookup: query_key type and MapArray key_type don't match. Expected "
+        "type: ",
+        *item_type, ", but got type: ", *options.query_key->type);
+  }
+
+  if (options.occurrence == MapArrayLookupOptions::Occurrence::ALL) {
+    return ValueDescr(list(item_type), descrs.front().shape);
+  } else /* occurrence == FIRST || LAST */ {
+    return ValueDescr(item_type, descrs.front().shape);
+  }
+}
+
+void AddMapArrayLookupKernels(ScalarFunction* func) {
+  for (const auto shape : {ValueDescr::ARRAY, ValueDescr::SCALAR}) {
+    ScalarKernel kernel({InputType(Type::MAP, shape)},
+                        OutputType(ResolveMapArrayLookupType),
+                        shape == ValueDescr::ARRAY ? MapArrayLookupFunctor::ExecMapArray
+                                                   : MapArrayLookupFunctor::ExecMapScalar,
```

Comment:
In order to use `VisitArrayValuesInline` this will have to change since you need to template `MapArrayLookupFunctor` with a `template <typename Type>`. Normally we would register one kernel for each supported type here, however, that breaks down because the input type needs to be parametric and we don't support that.

We could make a custom type mapper, however, instead, I think we can do the following and use a visitor

```
ScalarKernel kernel(InputType(Type::MAP), ..., ResolveMapArrayLookup::Exec);

struct ResolveMapArrayLookup {
  KernelContext* ctx;
  const ExecBatch& batch;
  Datum* out;

  template <typename Type>
  Status Execute() {
    if (batch[0].kind() == Datum::SCALAR) {
      return MapArrayLookupFunctor::ExecMapScalar(ctx, batch, out);
    }
    return MapArrayLookupFunctor::ExecMapArray(ctx, batch, out)
  }

  template <typename Type>
  enable_if_physical_integer<Type, Status> Visit(const Type& type) {
    return Execute<Type>();
  }

  // ...

  Status Visit(const DataType& type) { return Status::TypeError(...); }

  static Status Exec(...) {
    ResolveMapArrayLookup visitor{ctx, batch, out};
    return VisitTypeInline(checked_cast<const MapType&>(*batch[0].type()), *this);
  }
};
```
Another visitor example: https://github.com/apache/arrow/blob/91e3ac53e2e21736ce6291d73fc37da6fa21259d/cpp/src/arrow/compute/kernels/scalar_set_lookup.cc#L332

The custom type mapper would look something like this, I think:
https://github.com/apache/arrow/blob/8d0efd98bd4c4f2d181e01c02a66f924cb3e4c94/cpp/src/arrow/compute/kernel.cc#L176

You would define a custom TypeMatcher that checks the map key type, then register it as follows

```
for (const auto& type : IntTypes()) {
  ScalarKernel kernel({InputType(MapKeyTypeMatcher(type)), ..., GeneratePhysicalNumeric<MapArrayLookupFunctor>(type));
}
```

(I think I gave examples of this earlier)